File: core/src/main/java/de/qabel/core/repository/sqlite/AbstractSqliteRepository.java
package de.qabel.core.repository.sqlite;

import de.qabel.core.StringUtils;
import de.qabel.core.repository.exception.EntityNotFoundException;
import de.qabel.core.repository.exception.PersistenceException;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Collection;

public abstract class AbstractSqliteRepository<T> {
    protected ClientDatabase database;
    protected Hydrator<T> hydrator;
    protected String tableName;

    public AbstractSqliteRepository(ClientDatabase database, Hydrator<T> hydrator, String tableName) {
        this.database = database;
        this.hydrator = hydrator;
        this.tableName = tableName;
    }

    protected String getQueryPrefix() {
        return new StringBuilder("SELECT ")
            .append(StringUtils.join(", ", hydrator.getFields("t")))
            .append(" FROM ").append(tableName).append(" t ")
            .toString();
    }

    protected T findBy(String condition, Object... params) throws EntityNotFoundException, PersistenceException {
        String query = getQueryPrefix() + " WHERE " + condition + " LIMIT 1";
        return findByQuery(query, params);
    }

    protected T findByQuery(String query, Object[] params) throws EntityNotFoundException, PersistenceException {
        try (PreparedStatement statement = database.prepare(query)) {
            for (int i = 0; i < params.length; i++) {
                statement.setObject(i+1, params[i]);
            }
            try (ResultSet resultSet = statement.executeQuery()) {
                if (!resultSet.next()) {
                    throw new EntityNotFoundException("no entry for '" + query + "' and " + params + " -> " + params[0]);
                }
                return hydrator.hydrateOne(resultSet);
            }
        } catch (SQLException e) {
            throw new PersistenceException(
                "query failed: '" + query + "' and " + params + "(" + e.getMessage() + ")",
                e
            );
        }
    }

    protected Collection<T> findAll(String condition, Object... params) throws PersistenceException {
        String query = getQueryPrefix() + (condition.isEmpty() ? "" : " WHERE " + condition);
        String defaultOrder = getDefaultOrder();
        if (defaultOrder != null) {
            query += " ORDER BY " + defaultOrder;
        }
        try (PreparedStatement statement = database.prepare(query)) {
            for (int i = 0; i < params.length; i++) {
                statement.setObject(i+1, params[i]);
            }
            try (ResultSet resultSet = statement.executeQuery()) {
                return hydrator.hydrateAll(resultSet);
            }
        } catch (SQLException e) {
            throw new PersistenceException(
                "query failed: '" + query + "' (" + e.getMessage() + ")",
                e
            );
        }
    }

    protected String getDefaultOrder() {
        return null;
    }
}

File: core/src/main/java/de/qabel/core/repository/sqlite/AbstractSqliteRepository.kt
package de.qabel.core.repository.sqlite

import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException

import de.qabel.core.StringUtils
import de.qabel.core.repository.exception.EntityNotFoundException
import de.qabel.core.repository.exception.PersistenceException

abstract class AbstractSqliteRepository<T>(protected var database: ClientDatabase, protected var hydrator: Hydrator<T>, protected var tableName: String) {

    open protected val queryPrefix: String
        get() = StringBuilder("SELECT ").append(StringUtils.join(", ", hydrator.getFields("t"))).append(" FROM ").append(tableName).append(" t ").toString()

    @Throws(EntityNotFoundException::class, PersistenceException::class)
    protected fun findBy(condition: String, vararg params: Any): T {
        val query = "$queryPrefix WHERE $condition LIMIT 1"
        return findByQuery(query, params)
    }

    @Throws(EntityNotFoundException::class, PersistenceException::class)
    protected fun findByQuery(query: String, params: Array<out Any>): T {
        try {
            database.prepare(query).use { statement ->
                for (i in params.indices) {
                    statement.setObject(i + 1, params[i])
                }
                statement.executeQuery().use { resultSet ->
                    if (!resultSet.next()) {
                        throw EntityNotFoundException("no entry for '" + query + "' and " + params + " -> " + params[0])
                    }
                    return hydrator.hydrateOne(resultSet)
                }
            }
        } catch (e: SQLException) {
            throw PersistenceException(
                    "query failed: '" + query + "' and " + params + "(" + e.message + ")",
                    e)
        }

    }

    @Throws(PersistenceException::class)
    protected fun findAll(condition: String, vararg params: Any): Collection<T> {
        val query = queryPrefix + if (condition.isEmpty()) "" else " WHERE " + condition
        try {
            database.prepare(query).use { statement ->
                for (i in params.indices) {
                    statement.setObject(i + 1, params[i])
                }
                statement.executeQuery().use { resultSet -> return hydrator.hydrateAll(resultSet) }
            }
        } catch (e: SQLException) {
            throw PersistenceException(
                    "query failed: '" + query + "' (" + e.message + ")",
                    e)
        }

    }
}
